### ISSUE TITLE
workflow: population-based training / automated hyperparameter search lane

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#4019]
## Goal / Problem

Add a **population-based training (PBT) / automated hyperparameter search** lane to tune RL
training (learning rate, entropy, network/curriculum params) systematically instead of by hand,
improving reproducibility of the strongest baselines used in the benchmark.

## Scope

- In scope: a PBT or automated HPO harness over the existing training entry points; result/seed
  provenance capture; a small demonstration sweep.
- Out of scope: a managed-cluster scheduler redesign; benchmark claim before paper-grade.
- Entry points: training entry points, SLURM submission scripts, config conventions.

## Added Value / Effort / Complexity / Risk

- Value: medium (stronger, reproducible baselines). Effort: Low. Complexity: low–medium
  (orchestration). Risk: low (cost-aware; bound the sweep).

## Definition of Done

- [ ] PBT/HPO harness runs a bounded demonstration sweep with captured provenance.
- [ ] Best-config selection reproducible from recorded artifacts.
- [ ] Cost/scope bounds documented.

## Validation / Testing

- [ ] Small bounded sweep smoke run; repository validation gate.

## Related issues

- Supersedes old closed #430 (expert PPO hyperparameter grid). Relates: baseline lanes (Mamba/
  recurrent, this batch), training campaigns.

## Project Metadata

- Priority: low
- Effort (h): TBD
- Reviewed: No

**Comment by ll7** [comment]
## Implementation plan

### Design choice

Implement the first lane as **config-driven Optuna HPO** over the existing PPO training entry point, not as full mid-training PBT. This satisfies the issue scope (`PBT or automated HPO`) with lower scheduler and training-loop risk. Full PBT-style exploit/explore can be layered later if needed, but it should not be required for this issue’s DoD.

Current repository anchors to reuse:

- `scripts/training/train_ppo.py` remains the supported PPO expert training entry point.
- `scripts/training/optuna_expert_ppo.py` already runs Optuna trials against `run_expert_training()`.
- `scripts/training/launch_optuna_expert_ppo.py` already provides a config-first launcher.
- `configs/training/ppo_imitation/optuna_expert_ppo.yaml` is the existing launcher config surface.
- `scripts/tools/inspect_optuna_db.py` already summarizes and exports Optuna DB contents.
- `scripts/training/optuna_feature_extractor.py` contains a reusable pattern for optional SLURM fan-out.
- `docs/dev/slurm_submission.md` defines the current submission and traceability rules.
- Related closed issue #430 records the prior expert-PPO HPO history and the config-first gap.

### Phase 1 — Make HPO configuration explicit and reproducible

1. Keep `scripts/training/optuna_expert_ppo.py` as the low-level engine, but move the hard-coded search space behind a schema-backed config model.
2. Extend `scripts/training/launch_optuna_expert_ppo.py` to accept a new launcher schema, for example `robot_sf.optuna_expert_ppo_launcher.v2`, while preserving `v1` behavior.
3. Add a small search-space parser, preferably under `robot_sf/training/optuna_search_space.py`, with deterministic validation for:
   - `float` ranges with optional `log: true`,
   - `int` ranges or categorical integer values,
   - categorical values,
   - nested config patches.
4. Support these initial patch targets:
   - `ppo_hyperparams.learning_rate`, `ent_coef`, `n_steps`, `batch_size`, `n_epochs`, `clip_range`, `target_kl`, `gamma`, `gae_lambda`,
   - `policy_net_arch`,
   - `feature_extractor` and `feature_extractor_kwargs` where already supported by `ExpertTrainingConfig`,
   - a tightly bounded curriculum/config patch surface under existing `env_factory_kwargs` or `scenario_sampling` keys, only if the patch can be validated without arbitrary YAML mutation.
5. Preserve the existing default PPO search space when no explicit `search_space` block is provided, so older launch configs and notes remain usable.
6. Fix the known #430 gap while touching this surface: when `trial_timesteps` is omitted in the launcher/CLI, default to the base config’s `total_timesteps` rather than silently using an unrelated fixed default.

Example target config shape:

```yaml
schema_version: robot_sf.optuna_expert_ppo_launcher.v2
base_config: expert_ppo.yaml
trials: 4
metric: eval_episode_return
objective_mode: last_n_mean
objective_window: 2
trial_timesteps: 32000
eval_every: 16000
eval_episodes: 2
study_name: issue_4019_hpo_demo
storage: sqlite:///output/optuna/issue_4019_hpo_demo.db
seed: 4019
disable_wandb: true
deterministic: true
cost_bounds:
  max_trials: 4
  max_trial_timesteps: 32000
  max_eval_episodes: 2
  intended_evidence_tier: smoke
search_space:
  ppo_hyperparams:
    learning_rate: {type: float, low: 0.00001, high: 0.0003, log: true}
    ent_coef: {type: float, low: 0.0001, high: 0.02, log: true}
    n_steps: {type: categorical, values: [512, 1024]}
    batch_size: {type: categorical, values: [128, 256]}
  policy_net_arch:
    type: categorical
    values:
      - [64, 64]
      - [128, 128]
```

### Phase 2 — Capture trial provenance as first-class artifacts

Add an HPO-specific provenance writer, preferably `robot_sf/training/optuna_provenance.py`, called once per trial and once at study close.

Per-trial provenance should record at minimum:

- study name, storage URL with credentials redacted, Optuna trial number, trial state, objective value, objective metric/mode/window,
- base config path and SHA-256 hash,
- launcher config path and SHA-256 hash,
- resolved effective training config for the trial or a hash plus path to a rendered snapshot,
- search-space hash,
- sampler seed and deterministic flag,
- trial runtime bounds: `trial_timesteps`, `eval_every`, `eval_episodes`, number of seeds, resolved `num_envs`, worker mode,
- Git commit, branch, and dirty-tree flag when available,
- generated `policy_id`, checkpoint path, best checkpoint path, expert manifest path, training-run manifest path, episode log path, eval timeline path,
- safety constraint settings and feasibility result when configured,
- failure type/message for failed or pruned trials.

Implementation detail: continue storing the most important fields as Optuna `user_attrs`, but also write JSON/JSONL under an output root such as:

```text
output/optuna/<study_name>/
  study_manifest.json
  trials/trial_000.json
  trials/trial_001.json
  best_trial.json
  best_config.yaml
  selection_report.md
```

The Optuna DB should remain the authoritative study store; the JSON/YAML files make review and archival easier and satisfy the provenance requirement without requiring Optuna tooling for every downstream reader.

### Phase 3 — Reproducible best-config selection

Add a focused export tool, for example:

```bash
uv run python scripts/tools/export_optuna_best_config.py \
  --db output/optuna/issue_4019_hpo_demo.db \
  --study-name issue_4019_hpo_demo \
  --base-config configs/training/ppo_imitation/expert_ppo.yaml \
  --out output/optuna/issue_4019_hpo_demo/best_config.yaml \
  --report output/optuna/issue_4019_hpo_demo/selection_report.md
```

Selection rules:

1. Load the requested study from the recorded storage.
2. Filter to `COMPLETE` trials.
3. If safety constraints are active, prefer trials with `safety_constraints_feasible == true`.
4. Sort according to the study direction.
5. Reconstruct the effective config by applying the recorded best trial parameters to the recorded base config.
6. Write `best_config.yaml`, `best_trial.json`, and `selection_report.md` with hashes of all inputs.
7. Fail closed when required metadata is missing rather than silently guessing.

This makes the “best-config selection reproducible from recorded artifacts” DoD mechanically testable.

### Phase 4 — SLURM and cost-bounded execution surface

Do not redesign the scheduler. Add only a bounded, repository-local execution surface:

1. Add either `SLURM/issue_4019_optuna_hpo_demo.sl` or a planned `experiments/submission_queue.yaml` entry that runs the config-first launcher.
2. Use the existing `scripts/dev/sbatch_use_max_time.sh` wrapper where a long job is intended.
3. Keep the demo sweep local/sequential by default; use SLURM only as an optional execution route for GPU/cluster parity.
4. Document bounds in the launcher YAML and docs:
   - max trials,
   - max timesteps per trial,
   - eval cadence and eval episodes,
   - expected evidence tier (`smoke`, not benchmark evidence),
   - output root,
   - exact inspection/export commands.
5. If parallel HPO is needed, adapt the distributed Optuna pattern from `scripts/training/optuna_feature_extractor.py` in a minimal way: multiple workers run `--trials 1` against shared storage. Treat this as optional for #4019 unless the demo becomes too slow.

### Phase 5 — Demonstration sweep

Add a bounded demonstration config, for example:

```text
configs/training/ppo_imitation/optuna_expert_ppo_issue_4019_demo.yaml
```

Run a smoke-scale sweep with intentionally small cost:

```bash
uv run python scripts/training/launch_optuna_expert_ppo.py \
  --config configs/training/ppo_imitation/optuna_expert_ppo_issue_4019_demo.yaml \
  --disable-wandb \
  --deterministic \
  --log-level WARNING
```

Then inspect and export:

```bash
uv run python scripts/tools/inspect_optuna_db.py \
  --db output/optuna/issue_4019_hpo_demo.db \
  --study-name issue_4019_hpo_demo \
  --show-params \
  --top-n 4 \
  --export-csv output/optuna/issue_4019_hpo_demo/trials.csv

uv run python scripts/tools/export_optuna_best_config.py \
  --db output/optuna/issue_4019_hpo_demo.db \
  --study-name issue_4019_hpo_demo \
  --base-config configs/training/ppo_imitation/expert_ppo.yaml \
  --out output/optuna/issue_4019_hpo_demo/best_config.yaml \
  --report output/optuna/issue_4019_hpo_demo/selection_report.md
```

Attach the demo summary to `docs/training/` or `docs/context/` with an explicit claim boundary: it proves harness/provenance/selectability only; it does not establish a stronger benchmark baseline.

### Phase 6 — Tests and validation gates

Add targeted tests for the new behavior:

- search-space parser accepts valid float/int/categorical specs and rejects unknown mutation targets,
- trial config patching updates only the intended `ExpertTrainingConfig` fields,
- launcher `v1` behavior remains backward-compatible,
- launcher `v2` resolves `trial_timesteps` from the base config when omitted,
- provenance writer redacts storage credentials and records required hashes/paths,
- best-config exporter chooses the correct trial by optimization direction and safety feasibility,
- failed/pruned trials do not become selectable best configs.

Recommended focused validation:

```bash
uv run pytest \
  tests/training/test_optuna_objective.py \
  tests/training/test_optuna_expert_ppo_constraints.py \
  tests/training/test_optuna_search_space.py \
  tests/training/test_optuna_provenance.py \
  tests/tools/test_export_optuna_best_config.py \
  -q

uv run python scripts/training/launch_optuna_expert_ppo.py \
  --config configs/training/ppo_imitation/optuna_expert_ppo_issue_4019_demo.yaml \
  --dry-run \
  --log-level WARNING
```

Because this touches runtime behavior, metric semantics, and provenance, final PR validation should use the repository readiness gate:

```bash
BASE_REF=origin/main scripts/dev/pr_ready_check.sh
```

### Acceptance mapping

- **PBT/HPO harness runs a bounded demonstration sweep with captured provenance**: satisfied by the `v2` Optuna launcher config, per-trial/study manifests, and `issue_4019_hpo_demo` smoke sweep.
- **Best-config selection reproducible from recorded artifacts**: satisfied by `export_optuna_best_config.py`, `best_trial.json`, `best_config.yaml`, and `selection_report.md` with input hashes.
- **Cost/scope bounds documented**: satisfied by `cost_bounds` in the launcher config plus a docs page that records commands, limits, and the smoke-only claim boundary.
- **Small bounded sweep smoke run; repository validation gate**: satisfied by the demo command, focused tests, and `pr_ready_check.sh` before merge.

### Main risks

- Arbitrary nested YAML mutation can make trials irreproducible. Keep the first search-space schema allowlisted and explicit.
- SQLite-backed distributed Optuna can work for small cluster fan-out, but it is not a managed scheduler. Keep parallel execution optional and bounded.
- Smoke-scale results will be noisy. The demo should validate the harness only, not support benchmark or paper claims.
